File: ios/RoutineChart/RoutineChart/Core/Domain/UseCases/DeriveRoutineCompletionUseCase.swift
//
//  DeriveRoutineCompletionUseCase.swift
//  RoutineChart
//
//  Use case for deriving routine completion state
//

import Foundation

final class DeriveRoutineCompletionUseCase {
    private let stepRepository: RoutineStepRepository
    private let deriveStepCompletion: DeriveStepCompletionUseCase
    
    init(
        stepRepository: RoutineStepRepository,
        deriveStepCompletion: DeriveStepCompletionUseCase
    ) {
        self.stepRepository = stepRepository
        self.deriveStepCompletion = deriveStepCompletion
    }
    
    func execute(
        familyId: String,
        childId: String,
        routineId: String,
        localDayKey: String
    ) async throws -> Bool {
        // Get all steps for this routine
        let steps = try await stepRepository.getAll(routineId: routineId)
        
        // Routine is complete if ALL steps are complete
        for step in steps {
            let isStepComplete = try await deriveStepCompletion.execute(
                familyId: familyId,
                childId: childId,
                routineId: routineId,
                stepId: step.id,
                localDayKey: localDayKey
            )
            
            if !isStepComplete {
                return false
            }
        }
        
        return !steps.isEmpty // Empty routine is not complete
    }
}

File: android/app/src/main/java/com/HammersTech/RoutineChart/core/domain/usecases/DeriveRoutineCompletionUseCase.kt
package com.HammersTech.RoutineChart.core.domain.usecases

import com.HammersTech.RoutineChart.core.domain.models.CompletionRule
import com.HammersTech.RoutineChart.core.domain.models.Routine
import javax.inject.Inject

/**
 * Use case for deriving routine completion state
 * Checks if all steps meet the completion rule
 */
class DeriveRoutineCompletionUseCase @Inject constructor(
    private val deriveStepCompletionUseCase: DeriveStepCompletionUseCase
) {
    /**
     * Determine if a routine is complete for a given child and day
     * @return true if all steps are complete, false otherwise
     */
    suspend operator fun invoke(
        routine: Routine,
        childId: String,
        stepIds: List<String>,
        dayKey: String
    ): Boolean {
        return when (routine.completionRule) {
            CompletionRule.ALL_STEPS_REQUIRED -> {
                if (stepIds.isEmpty()) {
                    return false
                }

                val stepStates = deriveStepCompletionUseCase.getStepStates(
                    childId = childId,
                    routineId = routine.id,
                    stepIds = stepIds,
                    dayKey = dayKey
                )

                stepStates.values.all { it }
            }
        }
    }
}

